File: mvn-advent-of-code/src/main/java/advent/twenty_sixteen/Day05HowAboutANiceGameOfChess.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.common.Pair;
import advent.utilities.MessageDigestUtilities;

import java.security.NoSuchAlgorithmException;

public class Day05HowAboutANiceGameOfChess implements DailyProblem<String, String> {
    private final String part1Answer;
    private final String part2Answer;

    public Day05HowAboutANiceGameOfChess(String key) throws NoSuchAlgorithmException {
        MessageDigestUtilities mdUtils = new MessageDigestUtilities();

        String password = "";
        char[] password2 = new char[]{' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '};

        int count = 0;
        Pair<String, Integer> current = new Pair<>("", 9999);
        while (password.length() != 8 || !isValidPassword(password2)) {
            current = mdUtils.findNextHashWithXZeros(key, current.getSecond(), 5);

            // Part One
            if (password.length() != 8) {
                password += current.getFirst().charAt(5);
            }

            try {
                int index = Integer.parseInt("" + current.getFirst().charAt(5));
                char character = current.getFirst().charAt(6);
                if (index >= 0 && index < password2.length && password2[index] == ' ') {
                    password2[index] = character;
                }
            } catch (NumberFormatException e) {
            }
        }

        String temp = "";
        for (int i = 0; i != password2.length; i++) {
            temp += password2[i];
        }

        this.part1Answer = password.toLowerCase();
        this.part2Answer = temp.toLowerCase();
    }

    private boolean isValidPassword(char[] password) {
        int count = 0;
        for (int i = 0; i != password.length; i++) {
            if (Character.isLetterOrDigit(password[i])) {
                count++;
            }
        }
        return count == password.length;
    }

    @Override
    public String getPart1Answer() {
        return part1Answer;
    }

    @Override
    public String getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day05HowAboutANiceGameOfChess.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.ChecksumUtilities

import scala.annotation.tailrec

class Day05HowAboutANiceGameOfChess(input: String) extends DailyProblem[String, String] {

  private def generatePassword(password : Map[Char, Char]) : String = {
    "" +password('0') + password('1') + password('2') + password('3') + password('4') + password('5') + password('6') + password('7')
  }

  @tailrec
  private def generatePassword(index : Long, part1Password : String, part2Password : Map[Char, Char]) : (String, String) = {
    val result = generatePassword(part2Password)
    if (result.indexOf('-') < 0) {
      (part1Password, result)
    } else {
      val md5 = ChecksumUtilities.md5hashString(input + index)
      generatePassword(
        index + 1,
        if (md5.startsWith("00000") && part1Password.size < 8) part1Password + md5.charAt(5) else part1Password,
        if (md5.startsWith("00000") && part2Password(md5.charAt(5)) == '-') {
          part2Password + (md5.charAt(5) -> md5.charAt(6))
        } else {
          part2Password
        })
    }
  }

  val answer = generatePassword(0, "", Map[Char, Char]().withDefaultValue('-'))

  override val part1Answer: String = answer._1
  override val part2Answer: String = answer._2
}
